### PR TITLE
CEPH-11681: Tier-2 test to perform various operations on OMAPs and OSD OMAP compaction

### PR DESCRIPTION
**Pull request by harshkumarRH**
[CEPH-11681](https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-11681): Tier-2 test to perform various operations on OMAPs and OSD OMAP compaction

Test modules added/modified:
- `bench_write` in `ceph/rados/core_workflows.py`
- `perform_omap_operation` in `ceph/rados/pool_workflows.py`
- tests/rados/test_osd_compaction.py

Test suites modified:
- pacific/rados/tier-2_rados_test_omap.yaml
- quincy/rados/tier-2_rados_test_omap.yaml
- reef/rados/tier-2_rados_test_omap.yaml

Steps:
1. Create cluster with default configuration
2. Create a pool and write large number of OMAPS to few objects
3. Perform various operations on the OMAP entries, list, get,
    add new, delete, clear, etc
4. Trigger background IOPS using rados bench
5. Execute OSD compaction for a healthy OSD
6. Stop the OSD using systemctl
7. Execute OSD compaction on the down OSD, expected to fail

Logs:
Reef - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-CYJXY1/
Quincy - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-AQXX3Q/
Pacific - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-G58ST3/
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-ERRA0H

Signed-off-by: Harsh Kumar <hakumar@redhat.com>

### CODE REVIEW

**Comment by neha-gangadhar** [comment]
Looks good. Please add suite file where the test is intended to run and execution logs.

File: tests/rados/test_osd_compaction.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        )
+        log.info(
+            f"Printing last few OMAP keys and values present for Object {obj_rand}: "
+            f"{omap_kv_trim}"
+        )
+
+        # 3. create a custom omap key & value entry for an object
+        obj_rand = random.choice(omap_obj_list)
+        log.info(f"#3 Creating a custom OMAP key & value entry for obj: {obj_rand}")
+        custom_val = random.randint(pow(10, 5), pow(10, 6) - 1)
```
Comment (SrinivasaBharath) [inline tests/rados/test_osd_compaction.py:110]: Is there any logic to taking the random value in between 10^5 to 10^6?

File: tests/rados/test_osd_compaction.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        rados_obj.bench_write(
+            pool_name=bench_pool_name, rados_write_duration=300, background=True
+        )
+
+        # with IOPS in-progress perform OSD compaction of chosen OSD
+        log.info(f"Starting OSD compaction for OSD {osd_id}")
+        out, _ = cephadm.shell([f"ceph tell osd.{osd_id} compact"])
+        log.info(out)
+        assert "elapsed_time" in out
+        time.sleep(10)
```
Comment (SrinivasaBharath) [inline tests/rados/test_osd_compaction.py:184]: Later you can add one more check, If IOPS is not running then after the compaction the available memory of OSD should be greater than or equal to the previous memory.

Comment (harshkumarRH) [reply tests/rados/test_osd_compaction.py:222]: Added this check before running IOPS

**Comment by harshkumarRH** [comment]
> Looks good. Please add suite file where the test is intended to run and execution logs.

Added both, please review

File: tests/rados/test_osd_compaction.py (hunk 2 above)
Comment (harshkumarRH) [reply tests/rados/test_osd_compaction.py:221]: `Starting OSD compaction for OSD 13
2023-11-16 15:10:44,965 (cephci.test_osd_compaction) [INFO] - cephci.forked-repo-osd-compact.cephci.ceph.ceph.py:1557 - Running command cephadm -v shell -- ceph tell osd.13 compact on 10.0.208.53 timeout 600
2023-11-16 15:10:47,527 (cephci.test_osd_compaction) [INFO] - cephci.forked-repo-osd-compact.cephci.ceph.ceph.py:1591 - Command completed successfully
2023-11-16 15:10:47,529 (cephci.test_osd_compaction) [INFO] - cephci.forked-repo-osd-compact.cephci.tests.rados.test_osd_compaction.py:185 - {
    "elapsed_time": 0.33700545399999998
}

2023-11-16 15:10:57,534 (cephci.test_osd_compaction) [INFO] - cephci.forked-repo-osd-compact.cephci.ceph.ceph.py:1557 - Running command cephadm -v shell -- ceph osd df osd.13 -f json on 10.0.208.53 timeout 300
2023-11-16 15:10:59,659 (cephci.test_osd_compaction) [INFO] - cephci.forked-repo-osd-compact.cephci.ceph.ceph.py:1591 - Command completed successfully
2023-11-16 15:10:59,661 (cephci.test_osd_compaction) [INFO] - cephci.forked-repo-osd-compact.cephci.tests.rados.test_osd_compaction.py:194 - **Initial OSD utilization before compaction: 1.5901074630801688**
2023-11-16 15:10:59,662 (cephci.test_osd_compaction) [INFO] - cephci.forked-repo-osd-compact.cephci.tests.rados.test_osd_compaction.py:195 - **Final OSD utilization after compaction: 1.2836630967729332**`

**Review by SrinivasaBharath — APPROVED** [review]
LGTM

**Pull request merged**